File: src/src/lib/fileIcons.ts
import ArchiveIcon from '@fluentui/svg-icons/icons/folder_zip_20_filled.svg?no-inline'
import CodeIcon from '@fluentui/svg-icons/icons/code_20_filled.svg?no-inline'
import DocumentIcon from '@fluentui/svg-icons/icons/document_20_filled.svg?no-inline'
import FolderIcon from '@fluentui/svg-icons/icons/folder_20_filled.svg?no-inline'
import ImageIcon from '@fluentui/svg-icons/icons/image_20_filled.svg?no-inline'
import MusicIcon from '@fluentui/svg-icons/icons/music_note_2_20_filled.svg?no-inline'
import PdfIcon from '@fluentui/svg-icons/icons/document_pdf_20_filled.svg?no-inline'
import SlideIcon from '@fluentui/svg-icons/icons/slide_layout_20_filled.svg?no-inline'
import TableIcon from '@fluentui/svg-icons/icons/table_20_filled.svg?no-inline'
import TextIcon from '@fluentui/svg-icons/icons/document_text_20_filled.svg?no-inline'
import VideoIcon from '@fluentui/svg-icons/icons/video_20_filled.svg?no-inline'

export type FileIcon = {
  icon: string
  tone: string
  /** freedesktop icon name, used when an Omarchy icon theme is active. */
  themeName: string
}

const folderIcon: FileIcon = { icon: FolderIcon, tone: '#efa92a', themeName: 'folder' }
const genericIcon: FileIcon = { icon: DocumentIcon, tone: '#8a8a92', themeName: 'text-x-generic' }

const folderThemeNames: Record<string, string> = {
  documents: 'folder-documents', downloads: 'folder-download', music: 'folder-music',
  pictures: 'folder-pictures', videos: 'folder-videos', desktop: 'folder-desktop',
}

export function folderThemeName(folderName: string): string {
  return folderThemeNames[folderName.toLowerCase()] ?? 'folder'
}

const iconsByExtension: Record<string, FileIcon> = {}

function register(extensions: string[], icon: FileIcon) {
  for (const extension of extensions) {
    iconsByExtension[extension] = icon
  }
}

register(['png', 'jpg', 'jpeg', 'gif', 'webp', 'avif', 'bmp', 'svg', 'ico'], { icon: ImageIcon, tone: '#2b8fd6', themeName: 'image-x-generic' })
register(['mp3', 'wav', 'flac', 'm4a', 'ogg', 'opus', 'aac'], { icon: MusicIcon, tone: '#c94fb0', themeName: 'audio-x-generic' })
register(['mp4', 'mkv', 'mov', 'webm', 'avi', 'm4v'], { icon: VideoIcon, tone: '#e06c3b', themeName: 'video-x-generic' })
register(['doc', 'docx', 'odt', 'rtf'], { icon: DocumentIcon, tone: '#2b579a', themeName: 'x-office-document' })
register(['txt', 'md', 'log'], { icon: TextIcon, tone: '#6d6d75', themeName: 'text-x-generic' })
register(['xls', 'xlsx', 'ods', 'csv', 'tsv'], { icon: TableIcon, tone: '#217346', themeName: 'x-office-spreadsheet' })
register(['ppt', 'pptx', 'odp'], { icon: SlideIcon, tone: '#c43e1c', themeName: 'x-office-presentation' })
register(['pdf'], { icon: PdfIcon, tone: '#d13438', themeName: 'application-pdf' })
register(['zip', 'tar', 'gz', 'xz', 'zst', '7z', 'rar'], { icon: ArchiveIcon, tone: '#b58a1b', themeName: 'package-x-generic' })
register(
  ['ts', 'tsx', 'js', 'jsx', 'svelte', 'vue', 'rs', 'py', 'go', 'c', 'h', 'cpp', 'java', 'rb', 'sh', 'json', 'toml', 'yaml', 'yml', 'xml', 'xaml', 'html', 'css'],
  { icon: CodeIcon, tone: '#7a4fc0', themeName: 'text-x-script' },
)

export function fileIcon(name: string): FileIcon {
  const dotIndex = name.lastIndexOf('.')

  if (dotIndex <= 0) return folderIcon

  return iconsByExtension[name.slice(dotIndex + 1).toLowerCase()] ?? genericIcon
}

export { folderIcon }

File: core/src/file_icons.rs
//! Freedesktop icon names for files and folders, so every front end resolves
//! the same icon for the same entry.

const GENERIC: &str = "text-x-generic";

/// Standard folders get their own icon; everything else is a plain folder.
pub fn folder_icon_name(folder_name: &str) -> &'static str {
    match folder_name.to_ascii_lowercase().as_str() {
        "documents" => "folder-documents",
        "downloads" => "folder-download",
        "music" => "folder-music",
        "pictures" => "folder-pictures",
        "videos" => "folder-videos",
        "desktop" => "folder-desktop",
        _ => "folder",
    }
}

/// The icon for a file, chosen by extension.
///
/// A name with no extension, and a dotfile with nothing after its dot, both
/// fall back to the generic file icon — not to a folder, which is what the
/// Svelte build does for these.
pub fn file_icon_name(name: &str) -> &'static str {
    let Some(dot) = name.rfind('.') else { return GENERIC };
    if dot == 0 {
        return GENERIC;
    }
    match name[dot + 1..].to_ascii_lowercase().as_str() {
        "png" | "jpg" | "jpeg" | "gif" | "webp" | "avif" | "bmp" | "svg" | "ico" => "image-x-generic",
        "mp3" | "wav" | "flac" | "m4a" | "ogg" | "opus" | "aac" => "audio-x-generic",
        "mp4" | "mkv" | "mov" | "webm" | "avi" | "m4v" => "video-x-generic",
        "doc" | "docx" | "odt" | "rtf" => "x-office-document",
        "xls" | "xlsx" | "ods" | "csv" | "tsv" => "x-office-spreadsheet",
        "ppt" | "pptx" | "odp" => "x-office-presentation",
        "pdf" => "application-pdf",
        "zip" | "tar" | "gz" | "xz" | "zst" | "7z" | "rar" => "package-x-generic",
        "ts" | "tsx" | "js" | "jsx" | "svelte" | "vue" | "rs" | "py" | "go" | "c" | "h" | "cpp"
        | "java" | "rb" | "sh" | "json" | "toml" | "yaml" | "yml" | "xml" | "xaml" | "html" | "css" => "text-x-script",
        "txt" | "md" | "log" => GENERIC,
        _ => GENERIC,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extensions_choose_the_icon_and_case_does_not_matter() {
        assert_eq!(file_icon_name("holiday.JPEG"), "image-x-generic");
        assert_eq!(file_icon_name("report.pdf"), "application-pdf");
        assert_eq!(file_icon_name("main.rs"), "text-x-script");
        assert_eq!(file_icon_name("archive.tar.gz"), "package-x-generic");
        assert_eq!(file_icon_name("notes.unknown-extension"), GENERIC);
    }

    #[test]
    fn a_file_without_a_usable_extension_is_never_a_folder() {
        for name in ["Makefile", "LICENSE", ".bashrc", "trailing."] {
            assert_eq!(file_icon_name(name), GENERIC, "{name} should read as a file");
        }
    }

    #[test]
    fn standard_folders_keep_their_own_icons() {
        assert_eq!(folder_icon_name("Downloads"), "folder-download");
        assert_eq!(folder_icon_name("pictures"), "folder-pictures");
        assert_eq!(folder_icon_name("Code"), "folder");
    }
}
